src/script/interpreter.h

```diff
@@ -263,6 +263,8 @@ class BaseSignatureChecker
          return false;
     }
 
+    virtual bool CheckEqual(const std::vector<unsigned char>& vch1, const std::vector<unsigned char>& vch2) const { return vch1 == vch2; }
```

Comment:
I vaguely recall having a discussion about this already, but I can't remember the conclusion.

Is this function needed just to make the dummy signing pass verification in the signing logic? If so, it may be easier to only add a way to bypass that verification check for dummy signing in the first place.

I'm somewhat uncomfortable with making a change to consensus code like this just for that purpose.

Reply:
Yes, it's only to pass the `IsSolvable` check and, yes, it is unfortunate. Previous discussion [logs here](https://gnusha.org/miniscript/2022-02-24.log). TL;DR:
- I tested a few things and isolating a 3-lines patch to the interpreter was the only "reasonable" solution i could come up with that didn't change the existing logic
- `IsSolvable` is overkill already, we could just drop the `VerifyScript` there and not need the interpreter patch here:
https://github.com/bitcoin/bitcoin/blob/0897b189e4242321cd03339e7dde5c4d963d0e48/src/script/sign.cpp#L640-L646

But just getting rid of `VerifyScript` isn't enough. We still need to keep `ProduceSignature`, which will in turn call `VerifyScript` again. And since this path is used at "real" signing time too, it gets pretty convoluted to special case `IsSolvable` without using the `DummySignatureChecker`.

I think a cleaner path forward would be to revise the implementation of `IsSolvable` altogether, detangling it from the signing code. It dates from back when there were no descriptors. Nowadays everything we can sign for (that "is solvable") can be represented as a descriptor. So how about re-defining `IsSolvable` as "a descriptor can be inferred from this script, and this descriptor is solvable (ie not raw)"?
It's probably a good cleanup of legacy code on its own too, so it could have its own PR and i would rebase this one on top of it and drop the interpreter patch.

Reply:
IIRC, the DummySignatureCreator isn't just used for testing solvability, but also for determining the size of produced signatures before actual keys are available. If so, we still have a dependence on ProduceSignature in that context.

That said, it seems possible to just remove both VerifyScript calls in script/sign.cpp's ProduceSignature and IsSolvable. No tests break.

Reply:
Sorry for the late reply.

Yes. I think it would be nice to get rid of it entirely but indeed only redefining `IsSolvable` doesn't permit that, It's sufficient for dropping the interpreter patch here though.

I don't know about removing the belt-and-suspender  `VerifyScript` check in `ProduceSignature`. I'd be more comfortable with just re-defining `IsSolvable`. I've done that in #25664.